README


This archive contains the distribution FFI-Raw,
version 0.31:

  Perl bindings to the portable FFI library (libffi)

This software is copyright (c) 2012 by Alessandro Ghedini.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.


This README file was generated by Dist::Zilla::Plugin::Readme v5.013.



Changes
Revision history for FFI-Raw

0.31      2014-04-30 10:23:28+02:00 Europe/Rome

 - Re-upload to CPAN (was deleted by accident -.-")

0.30      2014-03-27 16:44:42+01:00 Europe/Rome

 - Fix build on Windows (GH#45)

0.29      2014-03-27 14:07:49+01:00 Europe/Rome

 - Avoid warnings when using undef to mean NULL (GH#44) (thanks, @plicease!)
 - Support returning pointers and strings from callbacks
   (GH#42) (GH#43) (thanks, @plicease!)
 - Update to libffi 3.1-rc1 (should hopefully fix a few build failures)

0.28      2014-02-14 13:09:10 Europe/Rome

 - Correctly handle unsigned values
 - Correctly handle return values on big endian platforms (GH#38)
 - Correctly handle absolute paths on Cygwin (GH#39) (thanks, @plicease!)

0.27      2014-02-03 19:44:11 Europe/Rome

 - Skip int64 tests if LLONG constants are not available
   (GH#35) (thanks, @plicease!)
 - Fix build on Cygwin64 (GH#37) (thanks, @plicease!)
 - Use ExtUtil::PkgConfig, if available, to detect system libffi
   (GH#36) (thanks, @plicease!)

0.26      2014-02-01 17:22:03 Europe/Rome

 - Import FFI::Raw::Ptr in FFI::Raw to avoid a bug in perl < 5.10.1
   (GH#23) (thanks, @plicease!)
 - Require perl >= 5.10
 - Add boolean override for FFI::Raw objects (GH#28) (thanks, @calid!)
 - Fix build on non-threaded perl (like OpenBSD's) (GH#29) (thanks, @plicease!)
 - Use system libffi if available (GH#30) (GH#32) (thanks, @plicease!)
 - Allow resolving functions in the main program (GH#33) (thanks, @plicease!)

0.25      2014-01-23 11:52:49 Europe/Rome

 - Fix build under MSWin32 64-bit (GH#17) (thanks, @plicease!)
 - Support explicit string length of 0 in MemPtr -> tostr()
   (GH#18) (thanks, @klaus)
 - Add FFI::Raw::Ptr base class for pointer objects
   (GH#20) (GH#22) (thanks, @plicease!)
 - Fix build under MS Visual C++ (GH#19) (thanks, @plicease!)

0.24      2013-12-19 16:44:50 Europe/Rome

 - Build libffi on MSWin32 (GH#15) (thanks, @plicease!)
 - Correctly translate undef to NULL and vice versa, when passing or returning
   a FFI::Raw::str or FFI::Raw::ptr (GH#16) (thanks, @plicease for the test!)

0.23      2013-12-04 20:45:24 Europe/Rome

 - Update libffi to 356b2cb
 - Fix some examples
 - (Try to )Fix a few build failures (again)

0.22      2013-12-01 17:51:57 Europe/Rome

 - Correctly handle callback coderef refcount (GH#14) (thanks, @plicease!)
 - Use ExtUtils::CBuilder for compiling C tests

0.21      2013-11-24 20:15:32 Europe/Rome

 - Add MemPtr -> new_from_ptr() (GH#12)
 - Fix a few build failures
 - Only recommend Math::Int64

0.20      2013-11-18 14:57:38 Europe/Rome

 - Fix warnings in the tests
 - Try to fix build failures on *BSD

0.19      2013-11-10 18:46:54 Europe/Rome

 - Revert libffi update (for some reason it makes the build fail sometimes)

0.18      2013-11-10 18:25:09 Europe/Rome

 - Add long and unsigned long ffi types (GH#7) (thanks, @calid!)
 - Add support for 64 bit ffi integer types (GH#9) (thanks, @calid!)
 - Fix invalid memory accesses in struct tests (should fix many failures)
 - Update libffi to 2f5b7ce

0.17      2013-03-02 19:20:35 Europe/Rome

 - Update libffi to 61a0549
 - Add FFI::Raw::MemPtr -> new_from_buf() method
 - Improve documentation for FFI::Raw::MemPtr and FFI::Raw::Callback
 - FFI::Raw::MemPtr -> tostr() method now takes an optional length parameter

0.16      2012-10-02 18:30:12 Europe/Rome

 - Fix build on OpenBSD

0.15      2012-09-30 14:35:47 Europe/Rome

 - Fix a number of test failures

0.14      2012-09-22 15:00:05 Europe/Rome

 - Fix build on Windows (GH#5)
 - Embed libffi sources to ease building

0.13      2012-09-19 18:44:55 Europe/Rome

 - Make types constants inlineable by perl (GH#2) (thanks, DOLMEN!)
 - Add a proper Changes file (GH#3)
 - Overload &{} to make a FFI::Raw object similar to a CODE ref (GH#4)

0.12      2012-08-20 11:34:16 Europe/Rome

 - Rebuild to update META.* repository links (no functional changes)

0.11      2012-08-14 12:07:18 Europe/Rome

 - Rebuild to remove internal test module from prereqs (no functional changes)

0.10      2012-08-14 11:40:47 Europe/Rome

 - Documentation fixes (no functional changes)
 - Code clean-up (no functional changes)

0.09      2012-07-19 09:36:28 Europe/Rome

 - Documentation fixes (no functional changes)
 - Remove experimental code notice (no functional changes)

0.08      2012-07-07 12:54:34 Europe/Rome

 - Rebuild to update META.* repository links (no functional changes)

0.07      2012-05-31 13:36:58 Europe/Rome

 - Make xt test suite the default (no functional changes)

0.06      2012-05-15 17:33:49 Europe/Rome

 - Add support for Perl callbacks
 - Add short and ushort types
 - Add experimental xt test suite
 - Fix libc path in tests (no functional changes)
 - Documentation fixes (no functional changes)

0.05      2012-05-09 11:32:43 Europe/Rome

 - Documentation fixes (no functional changes)
 - Add libuuid example
 - Add new_from_ptr() and ptr_to_str() subs
 - Use XS functions directly
 - Check for NULL library handle before closing it

0.04      2012-04-28 14:04:54 Europe/Rome

 - Fix libc test on Windows (no functional changes)
 - Fix FFI::Raw::MemPtr docs (no functional changes)

0.03      2012-04-26 16:35:05 Europe/Rome

 - Documentation improvements (no functional changes)
 - Add zeromq examples
 - Update malloc.pl example
 - Add FFI::Raw::MemPtr type and memptr() sub
 - Use perl's memory allocation macros

0.02      2012-04-19 15:12:42 Europe/Rome

 - Unload libraries on DESTROY
 - Fix argc check
 - Use a typemap
 - Check number of arguments in call()
 - Add uint and uchar types
 - Documentation fixes (no functional changes)

0.01      2012-04-18 15:36:21 Europe/Rome

 - Initial version


lib/FFI/Raw.pm
package FFI::Raw;
$FFI::Raw::VERSION = '0.31';
use strict;
use warnings;

require XSLoader;
XSLoader::load('FFI::Raw', $FFI::Raw::VERSION);

require FFI::Raw::Ptr;

use overload
	'&{}'  => \&coderef,
	'bool' => \&_bool;

sub _bool {
	my $ffi = shift;
	return $ffi;
}

=head1 NAME

FFI::Raw - Perl bindings to the portable FFI library (libffi)

=head1 VERSION

version 0.31

=head1 SYNOPSIS

    use FFI::Raw;

    my $cos = FFI::Raw -> new(
      'libm.so', 'cos',
      FFI::Raw::double, # return value
      FFI::Raw::double  # arg #1
    );

    say $cos -> call(2.0);

=head1 DESCRIPTION

B<FFI::Raw> provides a low-level foreign function interface (FFI) for Perl based
on L<libffi|http://sourceware.org/libffi/>. In essence, it can access and call
functions exported by shared libraries without the need to write C/XS code.

Dynamic symbols can be automatically resolved at runtime so that the only
information needed to use B<FFI::Raw> is the name (or path) of the target
library, the name of the function to call and its signature (though it is also
possible to pass a function pointer obtained, for example, using L<DynaLoader>).

Note that this module has nothing to do with L<FFI>.

=head1 METHODS

=head2 new( $library, $function, $return_type [, $arg_type ...] )

Create a new C<FFI::Raw> object. It loads C<$library>, finds the function
C<$function> with return type C<$return_type> and creates a calling interface.

If C<$library> is C<undef> then the function is searched in the main program.

This method also takes a variable number of types, representing the arguments
of the wanted function.

=head2 new_from_ptr( $function_ptr, $return_type [, $arg_type ...] )

Create a new C<FFI::Raw> object from the C<$function_ptr> function pointer.

This method also takes a variable number of types, representing the arguments
of the wanted function.

=head2 call( [$arg ...] )

Execute the C<FFI::Raw> function. This methoed also takes a variable number of
arguments, which are passed to the called function. The argument types must
match the types passed to C<new> (or C<new_from_ptr>).

The C<FFI::Raw> object can be used as a CODE reference as well. Dereferencing
the object will work just like call():

    $cos -> call(2.0); # normal call() call
    $cos -> (2.0);     # dereference as CODE ref

This works because FFI::Raw overloads the C<&{}> operator.

=head2 coderef( )

Return a code reference of a given C<FFI::Raw>.

=cut

sub coderef {
	my $ffi = shift;
	return sub { $ffi -> call(@_) };
}

=head1 SUBROUTINES

=head2 memptr( $length )

Create a L<FFI::Raw::MemPtr>. This is a shortcut for C<FFI::Raw::MemPtr-E<gt>new(...)>.

=cut

sub memptr { FFI::Raw::MemPtr -> new(@_) }

=head2 callback( $coderef, $ret_type [, $arg_type ...] )

Create a L<FFI::Raw::Callback>. This is a shortcut for C<FFI::Raw::Callback-E<gt>new(...)>.

=cut

sub callback { FFI::Raw::Callback -> new(@_) }

=head1 TYPES

=head2 FFI::Raw::void

Return a C<FFI::Raw> void type.

=cut

sub void ()  { ord 'v' }

=head2 FFI::Raw::int

Return a C<FFI::Raw> integer type.

=cut

sub int ()   { ord 'i' }

=head2 FFI::Raw::uint

Return a C<FFI::Raw> unsigned integer type.

=cut

sub uint ()   { ord 'I' }

=head2 FFI::Raw::short

Return a C<FFI::Raw> short integer type.

=cut

sub short ()   { ord 'z' }

=head2 FFI::Raw::ushort

Return a C<FFI::Raw> unsigned short integer type.

=cut

sub ushort ()   { ord 'Z' }

=head2 FFI::Raw::long

Return a C<FFI::Raw> long integer type.

=cut

sub long ()   { ord 'l' }

=head2 FFI::Raw::ulong

Return a C<FFI::Raw> unsigned long integer type.

=cut

sub ulong ()   { ord 'L' }

=head2 FFI::Raw::int64

Return a C<FFI::Raw> 64 bit integer type. This requires L<Math::Int64> to work.

=cut

sub int64 ()   { ord 'x' }

=head2 FFI::Raw::uint64

Return a C<FFI::Raw> unsigned 64 bit integer type. This requires L<Math::Int64> 
to work.

=cut

sub uint64 ()   { ord 'X' }

=head2 FFI::Raw::char

Return a C<FFI::Raw> char type.

=cut

sub char ()  { ord 'c' }

=head2 FFI::Raw::uchar

Return a C<FFI::Raw> unsigned char type.

=cut

sub uchar ()  { ord 'C' }

=head2 FFI::Raw::float

Return a C<FFI::Raw> float type.

=cut

sub float () { ord 'f' }

=head2 FFI::Raw::double

Return a C<FFI::Raw> double type.

=cut

sub double () { ord 'd' }

=head2 FFI::Raw::str

Return a C<FFI::Raw> string type.

=cut

sub str ()   { ord 's' }

=head2 FFI::Raw::ptr

Return a C<FFI::Raw> pointer type.

=cut

sub ptr ()   { ord 'p' }

=head1 AUTHOR

Alessandro Ghedini <alexbio@cpan.org>

=head1 SEE ALSO

L<FFI>, L<Ctypes|http://gitorious.org/perl-ctypes>

=head1 LICENSE AND COPYRIGHT

Copyright 2012 Alessandro Ghedini.

This program is free software; you can redistribute it and/or modify it
under the terms of either: the GNU General Public License as published
by the Free Software Foundation; or the Artistic License.

See http://dev.perl.org/licenses/ for more information.

=cut

1; # End of FFI::Raw


lib/FFI/Raw/Ptr.pm
package FFI::Raw::Ptr;
$FFI::Raw::Ptr::VERSION = '0.31';
use strict;
use warnings;

=head1 NAME

FFI::Raw::Ptr - Base FFI::Raw pointer type

=head1 VERSION

version 0.31

=head1 SYNOPSIS

    package Foo;

    use FFI::Raw;

    use base qw(FFI::Raw::Ptr);

    *_foo_new = FFI::Raw -> new(
      $shared, 'foo_new',
      FFI::Raw::ptr
    ) -> coderef;

    sub new {
      bless shift -> SUPER::new(_foo_new());
    }

    *get_bar = FFI::Raw -> new(
      $shared, 'foo_get_bar',
      FFI::Raw::int,
      FFI::Raw::ptr
    ) -> coderef;

    *set_bar = FFI::Raw -> new(
      $shared, 'foo_set_bar',
      FFI::Raw::void,
      FFI::Raw::ptr,
      FFI::Raw::int
    ) -> coderef;

    *DESTROY = FFI::Raw -> new(
      $shared, 'foo_free',
      FFI::Raw::void,
      FFI::Raw::ptr
    ) -> coderef;

    1;

    package main;

    my $foo = Foo -> new;

    $foo -> set_bar(42);

=head1 DESCRIPTION

A B<FFI::Raw::Ptr> represents a pointer to memory which can be passed to
functions taking a C<FFI::Raw::ptr> argument.

Note that differently from L<FFI::Raw::MemPtr>, C<FFI::Raw::Ptr> pointers are
not automatically deallocated once not in use anymore.

=head1 METHODS

=head2 new( $ptr )

Create a new C<FFI::Raw::Ptr> pointing to C<$ptr>, which can be either a
C<FFI::Raw::MemPtr> or a pointer returned by a C function.

=cut

sub new {
	my($class, $ptr) = @_;
	bless \$ptr, $class;
}

=head1 AUTHOR

Graham Ollis <plicease@cpan.org>

Alessandro Ghedini <alexbio@cpan.org>

=head1 LICENSE AND COPYRIGHT

Copyright 2014 Alessandro Ghedini.

This program is free software; you can redistribute it and/or modify it
under the terms of either: the GNU General Public License as published
by the Free Software Foundation; or the Artistic License.

See http://dev.perl.org/licenses/ for more information.

=cut

1; # End of FFI::Raw::Ptr


lib/FFI/Raw/MemPtr.pm
package FFI::Raw::MemPtr;
$FFI::Raw::MemPtr::VERSION = '0.31';
use strict;
use warnings;

=head1 NAME

FFI::Raw::MemPtr - FFI::Raw memory pointer type

=head1 VERSION

version 0.31

=head1 DESCRIPTION

A B<FFI::Raw::MemPtr> represents a memory pointer which can be passed to
functions taking a C<FFI::Raw::ptr> argument.

The allocated memory is automatically deallocated once the object is not in use
anymore.

=head1 METHODS

=head2 new( $length )

Allocate a new C<FFI::Raw::MemPtr> of size C<$length> bytes.

=head2 new_from_buf( $buffer, $length )

Allocate a new C<FFI::Raw::MemPtr> of size C<$length> bytes and copy C<$buffer>
into it. This can be used, for example, to pass a pointer to a function that
takes a C struct pointer, by using C<pack()> or the L<Convert::Binary::C> module
to create the actual struct content.

For example, consider the following C code

    struct some_struct {
      int some_int;
      char some_str[];
    };

    extern void take_one_struct(struct some_struct *arg) {
      if (arg -> some_int == 42)
        puts(arg -> some_str);
    }

It can be called using FFI::Raw as follows:

    use FFI::Raw;

    my $packed = pack('ix![p]p', 42, 'hello');
    my $arg = FFI::Raw::MemPtr -> new_from_buf($packed, length $packed);

    my $take_one_struct = FFI::Raw -> new(
      $shared, 'take_one_struct',
      FFI::Raw::void, FFI::Raw::ptr
    );

    $take_one_struct -> ($arg);

Which would print C<hello>.

=head2 new_from_ptr( $ptr )

Allocate a new C<FFI::Raw::MemPtr> pointing to the C<$ptr>, which can be either
a C<FFI::Raw::MemPtr> or a pointer returned by another function.

This is the C<FFI::Raw> equivalent of a pointer to a pointer.

=head2 tostr( [$length] )

Convert a C<FFI::Raw::MemPtr> to a Perl string. If C<$length> is not provided,
the length of the string will be computed using C<strlen()>.

=head1 AUTHOR

Alessandro Ghedini <alexbio@cpan.org>

=head1 LICENSE AND COPYRIGHT

Copyright 2013 Alessandro Ghedini.

This program is free software; you can redistribute it and/or modify it
under the terms of either: the GNU General Public License as published
by the Free Software Foundation; or the Artistic License.

See http://dev.perl.org/licenses/ for more information.

=cut

1; # End of FFI::Raw::MemPtr


lib/FFI/Raw/Callback.pm
package FFI::Raw::Callback;
$FFI::Raw::Callback::VERSION = '0.31';
use strict;
use warnings;

=head1 NAME

FFI::Raw::Callback - FFI::Raw function pointer type

=head1 VERSION

version 0.31

=head1 DESCRIPTION

A B<FFI::Raw::Callback> represents a function pointer to a Perl routine. It can
be passed to functions taking a C<FFI::Raw::ptr> type.

=head1 METHODS

=head2 new( $coderef, $ret_type [, $arg_type ...] )

Create a C<FFI::Raw::Callback> using the code reference C<$coderef> as body. The
signature (return and arguments types) must also be passed.

=head1 CAVEATS

For callbacks with a C<FFI::Raw::str> return type, the string value will be copied
to a private field on the callback object.  The memory for this value will be
freed the next time the callback is called, or when the callback itself is freed.
For more exact control over when the return value is freed, you can instead
use C<FFI::Raw::ptr> type and return a L<FFI::Raw::MemPtr> object.

=head1 AUTHOR

Alessandro Ghedini <alexbio@cpan.org>

=head1 LICENSE AND COPYRIGHT

Copyright 2013 Alessandro Ghedini.

This program is free software; you can redistribute it and/or modify it
under the terms of either: the GNU General Public License as published
by the Free Software Foundation; or the Artistic License.

See http://dev.perl.org/licenses/ for more information.

=cut

1; # End of FFI::Raw::Callback


Makefile.PL

use Config;
use Devel::CheckLib;

my $use_system_ffi = check_lib(lib => "ffi", header => "ffi.h");
my $pkg_config;

if (!$use_system_ffi && eval { require ExtUtils::PkgConfig }) {
  my %pkg_config = eval { ExtUtils::PkgConfig -> find('libffi') };

  unless ($@) {
    if (check_lib(header => "ffi.h", LIBS => $pkg_config{libs}, INC => $pkg_config{cflags})) {
      $use_system_ffi = 1;
      $pkg_config = \%pkg_config;
    }
  }
}

sub MY::postamble {
  return if $use_system_ffi;

  if ($^O eq 'MSWin32') {
    return "\t$^X -MAlien::MSYS=msys_run -Minc::MSYSConfigure -e \"configure(); msys_run 'make'\"\n\n";
  }

  return <<'MAKE_LIBFFI';
$(MYEXTLIB):
	cd deps/libffi && ./configure MAKEINFO=true --disable-builddir --with-pic && $(MAKE)

.NOTPARALLEL:

MAKE_LIBFFI
}

# This Makefile.PL for FFI-Raw was generated by
# inc::MakeMaker <self>
# and Dist::Zilla::Plugin::MakeMaker::Awesome 0.20.
# Don't edit it but the dist.ini and plugins used to construct it.

use strict;
use warnings;

use 5.010;

use ExtUtils::MakeMaker 6.30;



my %WriteMakefileArgs = (
  "ABSTRACT" => "Perl bindings to the portable FFI library (libffi)",
  "AUTHOR" => "Alessandro Ghedini <alexbio\@cpan.org>",
  "BUILD_REQUIRES" => {},
  "CONFIGURE_REQUIRES" => {
    "Devel::CheckLib" => 0,
    "ExtUtils::MakeMaker" => "6.30"
  },
  "DISTNAME" => "FFI-Raw",
  "EXE_FILES" => [],
  "INC" => "-I. -Ixs -Ideps -Ideps/libffi/include",
  "LICENSE" => "perl",
  "MYEXTLIB" => "deps/libffi/.libs/libffi.a",
  "NAME" => "FFI::Raw",
  "OBJECT" => "\$(O_FILES) deps/libffi/.libs/libffi.a",
  "PREREQ_PM" => {
    "XSLoader" => 0,
    "overload" => 0,
    "strict" => 0,
    "warnings" => 0
  },
  "TEST_REQUIRES" => {
    "Env" => 0,
    "ExtUtils::CBuilder" => 0,
    "File::Basename" => 0,
    "File::Copy" => 0,
    "File::Spec" => 0,
    "File::Temp" => 0,
    "FindBin" => 0,
    "IO::Handle" => 0,
    "IPC::Open3" => 0,
    "Math::BigInt" => 0,
    "POSIX" => 0,
    "Test::More" => 0,
    "Text::ParseWords" => 0,
    "base" => 0,
    "bigint" => 0,
    "lib" => 0
  },
  "VERSION" => "0.31",
  "test" => {
    "TESTS" => "t/*.t"
  }
);

my @libs;

$WriteMakefileArgs{CCFLAGS}   = "$Config{ccflags} $Config{cccdlflags}";
$WriteMakefileArgs{LDDLFLAGS} = "$Config{lddlflags}";

if ($^O eq 'MSWin32' && $Config{cc} =~ /cl(\.exe)?$/) {
  for (@WriteMakefileArgs{'MYEXTLIB','OBJECT'}) {
    s/libffi.a/libffi.lib/;
  }

  $WriteMakefileArgs{CCFLAGS}   .= " -DFFI_BUILDING",
  $WriteMakefileArgs{LDDLFLAGS} .= " psapi.lib";

} elsif ($^O =~ /^(MSWin32|cygwin)$/) {
  push @libs, '-L/usr/lib/w32api' if ($^O eq 'cygwin');
  push @libs, '-lpsapi';
}

if ($^O eq 'openbsd' && !$Config{usethreads}) {
  $WriteMakefileArgs{MYEXTLIB} .= ' /usr/lib/libpthread.a';
}

if ($use_system_ffi) {
  $WriteMakefileArgs{OBJECT} = '$(O_FILES)';

  if ($pkg_config) {
    push @libs, $pkg_config -> {libs};
    $WriteMakefileArgs{CCFLAGS} .= " " . $pkg_config -> {cflags};
  } else {
    push @libs, '-lffi';
  }

  delete $WriteMakefileArgs{MYEXTLIB};
}

$WriteMakefileArgs{LIBS} = "@libs" if @libs;



my %FallbackPrereqs = (
  "Env" => 0,
  "ExtUtils::CBuilder" => 0,
  "File::Basename" => 0,
  "File::Copy" => 0,
  "File::Spec" => 0,
  "File::Temp" => 0,
  "FindBin" => 0,
  "IO::Handle" => 0,
  "IPC::Open3" => 0,
  "Math::BigInt" => 0,
  "POSIX" => 0,
  "Test::More" => 0,
  "Text::ParseWords" => 0,
  "XSLoader" => 0,
  "base" => 0,
  "bigint" => 0,
  "lib" => 0,
  "overload" => 0,
  "strict" => 0,
  "warnings" => 0
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  delete $WriteMakefileArgs{TEST_REQUIRES};
  delete $WriteMakefileArgs{BUILD_REQUIRES};
  $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

if ( $^O eq 'MSWin32' ) {
	$WriteMakefileArgs{PREREQ_PM}{'Alien::MSYS'} = '0';
}

WriteMakefile(%WriteMakefileArgs);



